### PR TITLE
[Clang][Sema] Diagnose variable template explicit specializations with storage-class-specifiers

### PR DESCRIPTION
**Pull request by sdkrystian**
According to [[temp.expl.spec] p2](http://eel.is/c++draft/temp.expl.spec#2):
> The declaration in an _explicit-specialization_ shall not be an _export-declaration_. An explicit specialization shall not use a _storage-class-specifier_ other than `thread_local`.

Clang partially implements this, but a number of issues exist:
1. We don't diagnose class scope explicit specializations of variable templates with _storage-class-specifiers_, e.g.
    ```cpp
    struct A
    {
        template<typename T>
        static constexpr int x = 0;

        template<>
        static constexpr int x<void> = 1; // ill-formed, but clang accepts
    };
    ````
2. We incorrectly reject class scope explicit specializations of variable templates when `static` is not used, e.g.
    ```cpp
    struct A
    {
        template<typename T>
        static constexpr int x = 0;

        template<>
        constexpr int x<void> = 1; // error: non-static data member cannot be constexpr; did you intend to make it static?
    };
    ````
3. We don't diagnose dependent class scope explicit specializations of function templates with storage class specifiers, e.g.
    ```cpp
    template<typename T>
    struct A
    {
        template<typename U>
        static void f();

        template<>
        static void f<int>(); // ill-formed, but clang accepts
    };
    ````

This patch addresses these issues as follows:
- [#1]() is fixed by issuing a diagnostic when an explicit specialization of a variable template has storage class specifier
- [#2]() is fixed by considering any non-function declaration with any template parameter lists at class scope to be a static data member. This also allows for better error recovery (it's more likely the user intended to declare a variable template than a "field template").
- [#3]() is fixed by checking whether a function template explicit specialization has a storage class specifier even when the primary template is not yet known. 

One thing to note is that it would be far simpler to diagnose this when parsing the _decl-specifier-seq_, but such an implementation would necessitate a refactor of `ParsedTemplateInfo` which I believe to be outside the scope of this patch.

### CODE REVIEW

**Review by erichkeane — APPROVED** [review]
Sensible to me!  Definitely needs a release note.  Perhaps an alert on discourse about 'potentially breaking' changes.

File: clang/lib/Sema/SemaDeclCXX.cpp [hunk 1]
```diff
@@ -3561,6 +3562,7 @@ Sema::ActOnCXXMemberDeclarator(Scope *S, AccessSpecifier AS, Declarator &D,
       }
       return nullptr;
     }
+#endif
```
Comment (mizvekov) [inline clang/lib/Sema/SemaDeclCXX.cpp:3565]: Leftover

File: clang/lib/Sema/SemaDeclCXX.cpp [hunk 2]
```diff
@@ -3541,6 +3541,7 @@ Sema::ActOnCXXMemberDeclarator(Scope *S, AccessSpecifier AS, Declarator &D,
 
     IdentifierInfo *II = Name.getAsIdentifierInfo();
 
+#if 0
```
Comment (mizvekov) [inline clang/lib/Sema/SemaDeclCXX.cpp:3544]: Leftover

**Review by mizvekov — COMMENTED** [review]
Interestingly, only GCC errors on a case like this:

```
namespace {
  template<class T> int A = 0;
}
template<> int A<void> = 0;
```

https://godbolt.org/z/TTjssKxz5

**Review by Endilll — COMMENTED** [review]
DR test changes look good.

**Comment by sdkrystian** [comment]
@mizvekov GCC seems to be right; [[decl.meaning.general] p3.3](eel.is/c++draft/dcl.meaning.general#3.3) requires that the specialized declaration be nominable (per [[basic.scope.scope] p7](eel.is/c++draft/basic.scope.scope#7)) in `S` (which in this case is the global namespace). The primary template does not inhabit the global namespace nor any member of the inline namespace set thereof, so this should be ill-formed. Making the unnamed namespace declaration `inline` results in GCC accepting the declaration.

**Comment by mizvekov** [comment]
> @mizvekov GCC seems to be right; [[decl.meaning.general] p3.3](eel.is/c++draft/dcl.meaning.general#3.3) requires that the specialized declaration be nominable (per [[basic.scope.scope] p7](eel.is/c++draft/basic.scope.scope#7)) in `S` (which in this case is the global namespace). The target scope of the primary template is not the global namespace, nor is it any member of the inline namespace set thereof, so this should be ill-formed. Making the unnamed namespace declaration `inline` results in GCC accepting the declaration.

I think that's right. The GCC error message has bogus detail of course, there is no nested name specifier that would work here.
At least for clang, we are putting the specialized declaration in the anonymous namespace and giving it internal linkage, so this is not that bad.

I suppose fixing this is outside of the scope of this patch? If so, that would be alright.

**Comment by sdkrystian** [comment]
> I suppose fixing this is outside of the scope of this patch? If so, that would be alright.

@mizvekov  Yeah, definitely. I can submit a path to fix this either tomorrow or sometime next week.

File: clang/test/CXX/temp/temp.spec/temp.expl.spec/p2-20.cpp [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+static void D<int>::g(); // expected-warning {{explicit specialization cannot have a storage class}}
+                         // expected-error@-1 {{'static' can only be specified inside the class definition}}
+template<>
+template<>
+void D<int>::f<unsigned>();
+
+template<>
+template<>
+static void D<int>::g<unsigned>(); // expected-warning {{explicit specialization cannot have a storage class}}
+                                   // expected-error@-1 {{'static' can only be specified inside the class definition}}
```
Comment (mizvekov) [inline clang/test/CXX/temp/temp.spec/temp.expl.spec/p2-20.cpp:274]: It looks like this file doesn't have a trailing newline.

**Review by mizvekov — APPROVED** [review]
Small nit, otherwise LGTM, thanks!

**Comment by sdkrystian** [comment]
> > @mizvekov GCC seems to be right; [[decl.meaning.general] p3.3](eel.is/c++draft/dcl.meaning.general#3.3) requires that the specialized declaration be nominable (per [[basic.scope.scope] p7](eel.is/c++draft/basic.scope.scope#7)) in `S` (which in this case is the global namespace). The target scope of the primary template is not the global namespace, nor is it any member of the inline namespace set thereof, so this should be ill-formed. Making the unnamed namespace declaration `inline` results in GCC accepting the declaration.
> 
> I think that's right. The GCC error message has bogus detail of course, there is no nested name specifier that would work here. At least for clang, we are putting the specialized declaration in the anonymous namespace and giving it internal linkage, so this is not that bad.
> 
> I suppose fixing this is outside of the scope of this patch? If so, that would be alright.

@mizvekov Correction: GCC is wrong there. Embarrassingly, I'm the one who implemented clangs current behavior in [this patch](https://github.com/llvm/llvm-project/pull/80359) :P. The requirement for the specialized declaration to be nominable only applies to function template specializations.

**Comment by bgra8** [comment]
@sdkrystian we need a way to disable the new check so we can do the code fixes for our large codebase while still using the old compiler.

As the patch description suggests the correct code is rejected by the current `clang` so we have no path forward here. Can you please add a  flag to allow switching to the old behavior?

**Comment by alexfh** [comment]
Specifically, the problem is that we can't fix code that corresponds to case # 1 from the description, since it produces code corresponding to pattern # 2, which can't compile with current clang. Shipping fixes together with the compiler is not an option for us (and probably for many other users), so it should be possible to tell the new Clang to accept pattern # 1, i.e. it should be a warning (included into `-Wall`, probably) rather than a hard error for at least one compiler release.

**Comment by dwblaikie** [comment]
Sent a patch to add a warning flag for the warning this patch uses: https://github.com/llvm/llvm-project/pull/96699

With that, we could disable the warning during the compiler migration, decoupling compiler migration from code cleanup.

**Comment by alexfh** [comment]
There's more (https://gcc.godbolt.org/z/YKKah8M1n):
```
class D {
template <unsigned N>
static constexpr int kMaxUnsignedInt = 2 * kMaxUnsignedInt<N - 1> + 1;

template <>
constexpr auto kMaxUnsignedInt<1> = 1;
};
```
fails to compile with 
```
<source>:14:11: error: 'auto' not allowed in non-static class member
   14 | constexpr auto kMaxUnsignedInt<1> = 1;
      |           ^~~~
```

**Comment by sdkrystian** [comment]
> non-static class member

@alexfh Missed this one :) Opened a PR to fix this [here](https://github.com/llvm/llvm-project/pull/97425).

**Comment by alexfh** [comment]
Thanks for proposing a fix! I've been hinted though that Clang may be incorrect in accepting explicit specializations of a templated static class data member in the class scope. It looks like with all the restrictions in the standard it may only be allowed to place explicit specializations of a templated static class data member outside of the class. Though IANALL and I'm not sure what the _intention_ of the standard is in this area.

In any case, introducing any restrictions should be done with a clear migration path, so #97425 is likely helpful anyway.

**Comment by sdkrystian** [comment]
> I've been hinted though that Clang may be incorrect in accepting explicit specializations of a templated static class data member in the class scope. It looks like with all the restrictions in the standard it may only be allowed to place explicit specializations of a templated static class data member outside of the class.

@alexfh What makes you say that? I think the resolution of [CWG727](https://cplusplus.github.io/CWG/issues/727.html) quite clearly permits class scope explicit specializations of function templates, class templates, and static data member templates.

**Comment by dwblaikie** [comment]
> It looks like the presence of `static` on template variable specializations makes difference in the namespace context: https://gcc.godbolt.org/z/WGsreqbz8
> 
> Specifically, the specializations not marked `static` result in an exported variable. Thus, we have seemingly valid code that generates this diagnostic:
> 
> ```
> namespace A {
> template <unsigned N>
> static constexpr unsigned kMaxUnsignedInt = 2 * kMaxUnsignedInt<N - 1> + 1;
> 
> template <>
> static constexpr unsigned kMaxUnsignedInt<1> = 1;
> }
> ```
> 
> However, if we remove the `static` from the specialization, we're getting a potential ODR violation (considering this code can be in a header included into multiple translation units).
> 
> The right thing for this specific case is `inline constexpr`, but nevertheless, there seems to be an inconsistency now in how Clang handles the case now.

This seems to be pretty clearly a bug - in clang's specialization handling even prior to this patch.
GCC doesn't do this, it uses the storage class of the generic template for the specialization, which seems correct to me.

@sdkrystian are you interested in/able to fix this ^ ? https://gcc.godbolt.org/z/oPxKvvdeM

**Pull request merged**